### ISSUE TITLE
display and app

### ISSUE THREAD

**Issue by 123JackLin** [746505972/guandan#4]
"Just to confirm, when a human is playing (let's say, in a mode with a display or an app-like interface), is the setup that it's:
One human player, who receives suggestions from the trained AI model (the one loaded from the .pth file),
VS
Three AI opponents that are purely rule-based and untrained? (Meaning they understand the game rules and will make valid moves, but these moves are essentially chosen randomly or without the advanced strategy learned by the trained model.)"

**Comment by 746505972** [comment]
> Before you scroll down, please make sure you've updated to the latest version (I just uploaded it, which includes a batch of trained models) 🥰

This project supports **three display modes**:

1. `streamlit` front end

2. `Vue` front end
   Both of these use the `FrontendGame.py` to create a `GuandanGame` instance.

3. The third mode is via `display.py`, which runs in the terminal.
   In fact, you can interact with the game in **any file where a `GuandanGame` instance is created**.

In the older version, once you selected your seat and started the game, your opponents were bots that played cards randomly.
However, in the **latest version**, all your opponents are **trained AI models**.

Once you choose a `model_path` and assign yourself an AI suggestion model, your teammates and opponents will also use the same model.

You can modify the `play_turn` function inside the `GuandanGame` instance:

```python
    def play_turn(self):
        """Execute the current player's turn"""
        player = self.players[self.current_player]  # Get the current player object
        
        if self.user_player == self.current_player:
            result = self.user_play(player)
        else:  # If it's another player's turn, use actor_play to play
            result = self.actor_play(player)

        return result
```

You're free to customize this logic—for example, use `teammate` or `self.team` to make only your teammate use the AI model, while the opponents use the `ai_play()` random-bot strategy.

> Sorry for the misleading name `ai_play()`—in reality, `actor_play()` is the actual AI.
> By the way, I must say, the latest models are truly powerful—you should give them a try 🥰

---

>在您往下看之前，请先更新到最新版本（我刚刚上传的，加了一批训练好的模型）🥰

这个项目有3种展示方式，1.`streamlit`及2.`Vue`前端，他们都基于`FrontendGame.py`引入`GuandanGame`实例，

还有一种是3.`display.app`中在终端中交互，实际上，你可以在任何有`GuandanGame`实例的文件中交互。

在旧版本，当您选好位置并开始游戏后，您的对手确实是随机出牌的机器人，但在最新版本中，你的对手均为训练好的AI模型。

当您选择了`model_path`，为自己选择好了AI建议模型后，您的对手及队友同样来自这个模型。

您可以在`GuandanGame`实例的`play_turn`函数中修改。

```python
    def play_turn(self):
        """执行当前玩家的回合"""
        player = self.players[self.current_player]  # 获取当前玩家对象
        
        if self.user_player == self.current_player:
            result = self.user_play(player)
        else: # 当轮到其他座位上的玩家时，调用actor_play出牌
            result = self.actor_play(player)

        return result
```

您可以修改这里的逻辑，例如利用`teammate`或`self.team`等实现让队友同样是AI，而对手为`ai_play()`的随机机器人。

>抱歉我取了`ai_play()`这个让人误会的名字，实际上，`actor_play()`才是真正的AI
>另外，我不得不提，最新的模型真的很强大，您可以试试🥰

---
Here`s some running commands
### to run app.py
```bash
streamlit run app.py
```
### to run vue
1.start fastapi
```
guandan> uvicorn server.main:app --reload --port 8000
```
2.start vue(dev)

*in cmd*
```bash
cd guandan-frontend
npm run dev
```
you need to change the 3rd line in "guandan\guandan-frontend\src\utils\axios.ts" to fit your url.

**Comment by 123JackLin** [comment]
感谢作者大大的回答！另外我想问一下，在训练的过程中，智能体是一直跟人机（掌握规则，随机出牌）对打来训练的吗，也就是对手相对来说会比较菜吗？那这样训练出来的智能体的表现也会很好嘛，因为我是RL的小白，所以想问一下这个问题

**Comment by 746505972** [comment]
是的，A2C是一次只学习一方的行为，对局的质量对训练有影响。这样训练可能会导致模型学不到一下比较高阶的技巧（因为memery里就没出现过）。同时这种训练方式会导致模型学到针对它的对手策略，比如现在的人机在面对大牌时很轻易的就会把炸打出来，让模型学到了开局先出些大牌把他们的炸消耗完🥲。
有几种改进方法：

- 将人机的出牌策略修改，加上一些斗地主中成熟的算法，比如贪心等。这样可以将对局质量提高，让模型学得更好。
- 收集大量人类高质量对局然后编码给AI学习（这显然不是小项目能做到的，还需要和一些掼蛋游戏服务商联系取得数据）。

事实上我们采用transformer后胜率甚至能达到90%多，可以说达到目前的这个简陋环境能训练的最好效果了。

即使这样，最新的模型仍表现出了高于掼蛋新手的水准，而且具备一定的团队意识。

后续可以改进`ai_play()`的算法，或者改为A3C。

A3C就是4个actor对战，一局能采到4倍的样，但同时训练一局时间也翻4倍，而且可能会导致模型一直学不到某些打法，可能需要调整softmax的软硬程度。

**Comment by xzkeee** [comment]
加油加油

**Comment by hvtc-code** [comment]
作者大大，本人完全新手，想请教一下能不能在 ai_play() 中直接调用已有的训练结果作为出牌策略？另外，实际运用的时候和真人对战，怎么将对战过程记录下来给ai学习呢?

**Comment by 746505972** [comment]
第一个就是在play_turn()这里，把调用`ai_play()`改成`actor_play()`，这样那个位置上的人机就是用你选择的给建议的模型来出牌。
```py
    def play_turn(self):
        """执行当前玩家的回合"""
        player = self.players[self.current_player]  # 获取当前玩家对象
        
        if self.user_player == self.current_player:
            result = self.user_play(player)
        else: # 当轮到其他座位上的玩家时，调用actor_play出牌
            result = self.actor_play(player)

        return result
```
---

第二个我实际上没有做，不过鉴于整个环境都是自己搭的，你可以在每次轮到自己出牌的时候手动把`_get_obs()`等的内容保存下来，可以参考`actor.py`的210行之后的内容，把`memory`手动保存下来。之后再直接用记录的数据训练就行。
```py
state = game._get_obs()
memory:{"state": state, "action_id": action_id, "reward": reward}
```
如果想收集其他平台的数据就有点麻烦了，要么用CV要么抓包，然后再编制一次。

**Comment by hvtc-code** [comment]
比如我做如下更改行不行：将部分actor_play 中关于 `action `选择的部分添加到训练环境中的 `ai_play()`中
~~~python
state = self._get_obs()
        state_tensor = torch.tensor(state, dtype=torch.float32).unsqueeze(0)
        mask = torch.tensor(self.get_valid_action_mask(player.hand, M, self.active_level, self.last_play)).unsqueeze(0)
        probs = self.actor(state_tensor, mask)
        action_id = torch.multinomial(probs, 1).item()
        action_struct = M_id_dict[action_id]
~~~
这部分应该就是模型的推理过程了吧，如果我直接在`  possible_moves = self.get_possible_moves(player_hand) ` 后面，添加一个过滤条件，将possible_moves  中的action 识别出来，然后将不符合  `action_struct ` 建议的`type` 去掉，能不能简单的实现ai互博
抱歉格式稀烂，大佬见谅，我就想训练一个ai培长辈玩

**Comment by hvtc-code** [comment]
我不是计算机专业的，用github也不多，对于它的操作不大熟悉，见谅。
如果想实现了一部分 `逢人配` 规则，是不是主要修改`rule.py` `guandan_env.py` 以及 `get_actions.py` 中生成牌组的相关函数就行？
~~~python
 def _convert_wild_cards(self, cards):
        # print(f"debug: _convert_wild_cards:{cards}")
        """将万用牌转换为实际牌型"""

        # 判断卡牌组合内是否含有大小王，如果有，直接返回
        if any(card in ['大王', '小王'] for card in cards):
            return cards
        # 万能牌最多两张，后面需要添加判断，万能牌不能变成大小王
        wild_count = cards.count(self.wild_card)
        if wild_count == 0:
            return cards
        # 排除全是万能牌的可能
        if wild_count == len(cards):
            return cards
        # 复制牌组，排除万能牌
        converted = [card for card in cards if card != self.wild_card]

        # 获取非万用牌的点数和花色以及牌数量
        converted_ranks = [self.get_rank(card, as_one=False) for card in converted]
        converted_suits = [card[:2] for card in converted]

        # 依据牌组总长度判断可能的类型
        length = len(cards)

        if length == 1:
            # 单牌，万能牌作为普通级牌使用
            converted.append(self.wild_card)

        elif length == 2 and wild_count == 1:
            # 对子，万能牌变成另一张牌的点数
            other_card = converted[0]
            converted.append(other_card)

        elif length == 3:
            # 三同张，依据万能牌数量决定变化
            if wild_count == 2:
                # 2张万能牌，都变成普通牌的点数
                other_card = converted[0]
                converted.extend([other_card] * wild_count)
            elif wild_count == 1:
                # 1张万用牌，检查其他两张牌的点数是否一致
                if len(converted) == 2 and converted_ranks[0] == converted_ranks[1]:
                    converted.append(converted[0])
                else:
                    # 不合法的出牌，返回原牌组（实际应该报错）
                    return cards

        elif length == 4:
            # 4炸，依据万能牌数量决定变化
            if wild_count == 2:
                # 2张万用牌，检查其他2张牌的点数是否一致
                if len(converted) == 2 and converted_ranks[0] == converted_ranks[1]:
                    converted.extend([converted[0]] * wild_count)
                else:
                    return cards
            elif wild_count == 1:
                # 1张万用牌，检查其他3张牌的点数是否一致
                if len(converted) == 3 and len(set(converted_ranks)) == 1:
                    converted.append(converted[0])
                else:
                    return cards
~~~
大佬觉得这个思路可行吗？

**Comment by 746505972** [comment]
我这个项目现在已经有点屎山了，在前期我的手牌编码都是直接用`list[str]`中文花色+数字表示的，`rule.py`的识别也是基于字符串的处理的。而在后期需要训练模型，定义动作空间的时候我又改了一套输出，是`dict`结构的JSON，然后每次遇到2方出牌既有人机又有神经网络的时候就需要将一方的牌转换。

如果真要加逢人配我建议统一写成JSON结构，用`{"type":str,"points":list,"logic_point":int,"id":int}`这样的结构存，这时候只需要枚举，将points中的任意1~2个点数用`15（级牌）`代替。这样加起来也不多，应该不超过2000种结构。不过这时动作空间可能太大了，可以再加一个字段`if_wild_cards`表明是否用了逢人配，这样数量只需要翻倍甚至不变（之后要解码）就能做到。

也可以在发牌的之后直接给牌编上索引，总共就108张牌，之后直接用1-108的数字代替牌，这样写逢人配也比较不错。

其实我之前是写过逢人配的，但那样模型不好训练就砍掉了

**Comment by 746505972** [comment]
> 比如我做如下更改行不行：将部分actor_play 中关于 `action `选择的部分添加到训练环境中的 `ai_play()`中
> ```py
> state = self._get_obs()
>         state_tensor = torch.tensor(state, dtype=torch.float32).unsqueeze(0)
>         mask = torch.tensor(self.get_valid_action_mask(player.hand, M, self.active_level, self.last_play)).unsqueeze(0)
>         probs = self.actor(state_tensor, mask)
>         action_id = torch.multinomial(probs, 1).item()
>         action_struct = M_id_dict[action_id]
>```
> 这部分应该就是模型的推理过程了吧，如果我直接在` possible_moves = self.get_possible_moves(player_hand)` 后面，添加一个过滤条件，将possible_moves 中的action 识别出来，然后将不符合 `action_struct ` 建议的`type` 去掉，能不能简单的实现ai互博 抱歉格式稀烂，大佬见谅，我就想训练一个ai培长辈玩

还有这条有点没看懂想要干嘛

**Comment by hvtc-code** [comment]
> > 比如我做如下更改行不行：将部分actor_play 中关于 `action `选择的部分添加到训练环境中的 `ai_play()`中

> 
> 还有这条有点没看懂想要干嘛

在训练的时候，其中一个是 A2C 模型推理后出的结果`actor_play`，另外三个也就是是随机出牌，没有推理`ai_play`；我想在训练的时候把另外3个的出牌逻辑做调整，也能够实现推理后出牌。直接修改 `ai_play`的出牌逻辑能够实现吗？

**Comment by 746505972** [comment]
直接把`actor.py`的250行改成`game.actor_play(player)`即可。但是要注意它调用的模型，你可以复制一个新函数，它还接收一个参数，用来指定使用的模型。

另外，A3C我训练过，除了训练速度变慢以外，效果并没多好。也许是训练的不够久，也许是当前模型的上限就在这里了（毕竟这个模型实在很小，也简单），也可以调一下softmax的软硬程度

**Comment by hvtc-code** [comment]
好的，谢谢大佬指点。这几个模型有什么区别啊
`base.pth`  `master.pth` `noob.pth` `normal.pth` `overfit.pth`
我该怎么简单的和长辈解释需要选哪个模型来对局

**Comment by 746505972** [comment]
就单纯的训练轮数不同，分别对应`基础（200轮，基本只掌握规则）`,`大师`，`菜鸟（和基础不相上下）`，`普通`，`过拟合（走火入魔的大师？）`

**Comment by hvtc-code** [comment]
那和这些有什么区别？
`critic_ep1000.pth`  `actor_ep1000.pth`
比如我现在 想用你训练了3000 轮的 无上王者(手动狗头) ，该怎么设置

**Comment by 746505972** [comment]
`critic`是打分的模型，不能出牌。如果你用的是`streamlit`，修改`app.py`第33行为
```py
available_models = [f for f in os.listdir(model_dir) if f.endswith(".pth") and not f.startswith("c")]
```
vue类似，改`server/main.py`的第28行。

实际上前面那些只是给`actor_epxxx.pth`取了个名字，`master`好像就是2400轮的actor

**Comment by hvtc-code** [comment]
明白了，感谢大佬指点。

**Comment by hvtc-code** [comment]
> 第二个我实际上没有做，不过鉴于整个环境都是自己搭的，你可以在每次轮到自己出牌的时候手动把`_get_obs()`等的内容保存下来，可以参考`actor.py`的210行之后的内容，把`memory`手动保存下来。之后再直接用记录的数据训练就行。
> 
> state = game._get_obs()
> memory:{"state": state, "action_id": action_id, "reward": reward}
> 如果想收集其他平台的数据就有点麻烦了，要么用CV要么抓包，然后再编制一次。

大佬，我又来问问题了，`backend.py`  `FrontendGame.py`  这两个和 `guandan_env.py` 有啥区别吗？我看代码差不了太多？我想把用户出牌记录保存下来，应该修改那个文件？可以在哪添加你前面说的代码？

**Comment by 746505972** [comment]
没有区别，`FrontendGame.py`是实际在用的，`backend.py`是单文件版的，没有import其他文件，`guandan_env.py`这些是历史遗留的，只是方便测试。

应该在`FrontendGame.py`里改

**Comment by hvtc-code** [comment]
大佬，我看在`actor.py`以及 `A3C.py`中引用是这样的：
~~~python
from guandan_env import GuandanGame
from get_actions import enumerate_colorful_actions
~~~
意思是不是训练的时候用的是`guandan_env.py`，然后在 实际ai出牌的时候 用的是 `FrontendGame.py` ，那么该是不是可以手动把用户出牌信息也就是 memory  保存下来，然后在训练调用 `train_on_batch(memory)`，让AI学会用户出牌策略，改变现在先把大牌都出了，然后过单牌的逻辑

**Comment by 746505972** [comment]
是的，`guandan_env.py`是训练时用的环境，它没有import `torch`，`FrontendGame.py`是后端，它会调用训练好的模型，前端都是从它这调api的。

手动构建训练数据可能需要很长时间，至少需要几十把高质量对局。
